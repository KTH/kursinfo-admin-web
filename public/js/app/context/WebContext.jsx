import React from 'react'
import { addClientFunctionsToWebContext } from '../client-context/addClientFunctionsToWebContext'

const WebContext = React.createContext()

const defaultConfig = () => ({
  lang: 'sv',
  isAdmin: false,
  basicBreadcrumbs: [
    { label: 'KTH', url: 'https://www-r.referens.sys.kth.se/' },
    { label: 'Node', url: 'https://www-r.referens.sys.kth.se/node' },
  ],
  proxyPrefixPath: { uri: 'node' },
  message: 'howdi',
})

export const WebContextProvider = props => {
  const { configIn = {} } = props
  const config = { ...defaultConfig() }
  for (const k in configIn) {
    if (Object.prototype.hasOwnProperty.call(configIn, k)) {
      if (typeof configIn[k] === 'object') {
        config[k] = JSON.parse(JSON.stringify(configIn[k]))
      } else {
        config[k] = configIn[k]
      }
    }
  }

  // OBS! deviation from NODE-WEB to make functions working
  const [currentConfig, setConfig] = React.useState({ ...config, ...addClientFunctionsToWebContext() })
  const value = [currentConfig, setConfig]
  // eslint-disable-next-line react/jsx-props-no-spreading
  return <WebContext.Provider value={value} {...props} />
}

export function useWebContext() {
  const context = React.useContext(WebContext)
  if (!context) {
    throw new Error('useWebContext must be used within a WebContext')
  }
  return context
}
