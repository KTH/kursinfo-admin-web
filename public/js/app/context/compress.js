/* eslint no-use-before-define: ["error", "nofunc"] */

export { compressData, uncompressData }

/**
 * During server-side rendering put the result of this function
 * into <script>-tags so that it gets executed when the page is loaded.
 *
 * @param {data} data
 *
 * @returns {string}
 */
function compressData(data, dataId = 'DATA') {
  const compressedData = encodeURIComponent(JSON.stringify(data))

  const output = `window.__compressedData__${dataId} = "${compressedData}";`
  return output
}

/**
 * During client startup use this function to initialize the client
 * with the server-side prepared data
 *
 * @param {data} data
 *    Data to be used when populating context object
 *    which was created during server-side rendering
 *
 * @return {object}
 *    Returns same object as input
 *    but it might have been changed in place
 */
function uncompressData(data, dataId = 'DATA') {
  const isClientSide = typeof window !== 'undefined'
  if (!isClientSide) {
    // eslint-disable-next-line no-console
    console.error('uncompressData(): Expected to be run on client side')
    return data
  }

  const dataFound = window[`__compressedData__${dataId}`]
  const hasCompressedData = dataFound != null && typeof dataFound === 'string'
  if (!hasCompressedData) {
    return data
  }

  const uncompressedData = JSON.parse(decodeURIComponent(dataFound))
  const validData = uncompressedData != null && typeof uncompressedData === 'object'
  if (!validData) {
    // eslint-disable-next-line no-console
    console.error('uncompress(): Invalid  data')
    return data
  }

  const dataKeys = Object.keys(uncompressedData)
  dataKeys.forEach(key => {
    // eslint-disable-next-line no-param-reassign
    data[key] = uncompressedData[key]
  })
  return data
}
