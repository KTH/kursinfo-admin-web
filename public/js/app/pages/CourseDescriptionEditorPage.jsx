import React, { Component } from 'react'
import { inject, observer} from 'mobx-react'
import i18n from '../../../../i18n'

import CourseTitle from '../components/CourseTitle'
import ProgressBar from '../components/ProgressBar'
import PictureUpload from './PictureUpload'
import SellingInfo from './SellingInfo'
import Preview from '../components/PreviewText'


@inject(['adminStore']) @observer
class CourseDescriptionEditorPage extends Component {
  constructor (props) {
    super(props)
    this.state = {
      progress: props.progress ? Number(props.progress) : 1
    }
    this.koppsData = this.props.adminStore.koppsData
    this.storageUri = this.props.adminStore.browserConfig.storageUri
    this.langIndex = this.koppsData.lang === 'en' ? 0 : 1
    this.doUpdateStates = this.doUpdateStates.bind(this)
  }

  doUpdateStates (states) {
    if (states) this.setState(states)
  }

  render () {
    const { koppsData, langIndex } = this
    const { introLabel } = i18n.messages[langIndex]

    const { courseImage } = i18n.messages[1]    
    let courseImageID = courseImage[koppsData.mainSubject]
    if (courseImageID === undefined) courseImageID = courseImage.default
    const defaultImageUrl = `${this.storageUri}${courseImageID}`
    const introText = this.state.progress === 1 ? introLabel.step_1_desc :
                      this.state.progress === 2 ? introLabel.step_2_desc : introLabel.step_3_desc

    return (
      <div key='kursinfo-container' className='kursinfo-main-page col'>
        <CourseTitle key='title'
          courseTitleData={koppsData.courseTitleData}
          pageTitle={introLabel.editCourseIntro}
          language={koppsData.lang}
          />
        <ProgressBar active={this.state.progress} language={langIndex} introText={introText}/>
        {this.state.progress === 1
        ? <PictureUpload defaultImageUrl={defaultImageUrl} introLabel={introLabel}
          koppsData={koppsData}
          updateParent={this.doUpdateStates} />
        : this.state.progress === 2
          ? <SellingInfo koppsData={koppsData} updateParent={this.doUpdateStates}
          />
          : <Preview introLabel={introLabel} defaultImageUrl={defaultImageUrl}
            updateParent={this.doUpdateStates}
          />
        }
      </div>
    )
  }

}

export default CourseDescriptionEditorPage
