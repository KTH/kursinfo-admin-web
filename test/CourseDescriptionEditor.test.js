import React from 'react'
import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import '@babel/runtime/regenerator'
import mockWebContext from './mocks/mockWebContext'
import { mockClientFunctionsToWebContext } from './mocks/mockClientFunctionsToWebContext'
import CourseDescriptionEditorPage from '../public/js/app/pages/CourseDescriptionEditorPage'
import { WebContextProvider } from '../public/js/app/context/WebContext'

jest.mock('../public/js/app/client-context/addClientFunctionsToWebContext')

const renderEditPage = (newWebContext = {}, pageNumber) => {
  return render(
    <WebContextProvider
      configIn={{
        ...mockWebContext,
        ...newWebContext,
      }}
    >
      <CourseDescriptionEditorPage progress={pageNumber} />
    </WebContextProvider>
  )
}

const renderWithState = (stateToSet = {}, pageNumber) => {
  const newWebContext = Object.assign(Object.assign({}, mockWebContext), stateToSet)
  return renderEditPage(newWebContext, pageNumber)
}

describe('<CourseDescriptionEditorPage> (and subordinates)', () => {
  beforeAll(() => mockClientFunctionsToWebContext())
  afterAll(() => {
    jest.clearAllMocks()
  })

  test('Has correct main heading', () => {
    renderEditPage()
    const allH1Headers = screen.getAllByRole('heading', { level: 1 })
    expect(allH1Headers.length).toBe(1)
    expect(allH1Headers[0]).toHaveTextContent(/^Redigera introduktion till kursenSF1624 Algebra och geometri 7,5 hp/)
  })

  describe('Page 1A Välj Bild', () => {
    beforeAll(() => mockClientFunctionsToWebContext())
    afterAll(() => {
      jest.clearAllMocks()
    })

    test('Has correct name in progress bar', () => {
      renderEditPage().getByText('1. Välj bild')
    })

    test('Has correct heading', () => {
      const heading = renderEditPage().getByTestId('intro-heading')
      expect(heading).toHaveTextContent(/^Välj bild$/) // exact match
    })

    test('Has correct introductory text', () => {
      const introText = renderEditPage().getByTestId('intro-text')
      expect(introText).toHaveTextContent(
        'I steg 1 av 3 väljer du en bild att visa på sidan Inför kursval. I steg 2 av 3 lägger du in eller redigerar den inledande texten. I steg 3 av 3 granskar du bild och text för att sedan publicera.'
      )
    })
  })

  describe('Page 1B Ingen bild vald', () => {})

  const PUBLISHED_IMAGE_EXISTS = {
    isDefaultChosen: false,
    isApiPicAvailable: true,
    tempImagePath: undefined,
  }

  const IMAGE_SELECTED_FOR_UPLOAD = {
    isDefaultChosen: false,
    isApiPicAvailable: false,
    tempImagePath: 'ImageThatWasSelectedForUpload.png',
  }

  const OVERWRITE_PUBLISHED_IMAGE = {
    isDefaultChosen: false,
    isApiPicAvailable: true,
    tempImagePath: 'ImageThatWasSelectedForUpload.png',
  }

  describe('Page 1C Bildval felaktigt', () => {
    beforeAll(() => mockClientFunctionsToWebContext())
    afterAll(() => {
      jest.clearAllMocks()
    })
    const expectedAlert =
      'Du behöver välja en bild med rätt format (se markering i rött nedan) för att kunna gå vidare till ”Redigera text”.'
    const expectedErrorMessage = 'Obligatoriskt (format: .png eller .jpg)'

    test('Has correct alert text and error message (no image selected)', async () => {
      const { getByText, getByLabelText, getByRole, getByTestId, queryByTestId } = renderEditPage()

      expect(getByLabelText('Standardbild utifrån kursens huvudområde').checked).toBeTruthy()
      expect(getByLabelText('Egen bild').checked).toBeFalsy()
      getByLabelText('Egen bild').click()

      //No error message visible initially
      expect(queryByTestId('error-text')).toBeFalsy()

      getByText('Redigera text').click()

      //Alert and Error Message assertion
      expect(getByRole('alert')).toHaveTextContent(expectedAlert)
      const errorMessageSpan = getByTestId('error-text')
      expect(errorMessageSpan).toHaveTextContent(expectedErrorMessage)
    })

    test('Has correct alert text and error message (wrong format chosen)', () => {
      const { getByLabelText, getByRole, getByTestId, queryByTestId } = renderEditPage()

      expect(getByLabelText('Standardbild utifrån kursens huvudområde').checked).toBeTruthy()
      expect(getByLabelText('Egen bild').checked).toBeFalsy()
      getByLabelText('Egen bild').click()

      //No error message visible initially
      expect(queryByTestId('error-text')).toBeFalsy()

      //simulate file upload with incorrect format
      const imageInput = getByTestId('fileUpload')
      fireEvent.change(imageInput, {
        target: {
          files: [new File(['(⌐□_□)'], 'empty.txt', { type: 'text' })],
        },
      })

      //Alert and Error Message assertion
      expect(getByRole('alert')).toHaveTextContent(expectedAlert)
      const errorMessageSpan = getByTestId('error-text')
      expect(errorMessageSpan).toHaveTextContent(expectedErrorMessage)
    })
  })

  describe('Page 1D Första egen bild valt', () => {})

  describe('Page 1E Tidigare val av egen bild', () => {})

  describe('Page 1F Valt om till, ytterligare ny egen bild', () => {})

  describe('Page 1G Valt om, ämnesbild', () => {
    beforeAll(() => mockClientFunctionsToWebContext())
    afterAll(() => {
      jest.clearAllMocks()
    })

    const useDefaultImage = 'Standardbild utifrån kursens huvudområde'
    const expected =
      'Observera att den egna valda bilden som nu är publicerad kommer att raderas när du publicerar i steg 3.'

    test('Has correct alert text (published image, switch to default)', () => {
      const { getByRole, getByLabelText } = renderWithState(PUBLISHED_IMAGE_EXISTS)
      getByLabelText(useDefaultImage).click()
      expect(getByRole('alert')).toHaveTextContent(expected)
    })

    test('No alert (no published image, upload image, switch to default)', () => {
      const { queryByRole, getByLabelText } = renderWithState(IMAGE_SELECTED_FOR_UPLOAD)
      getByLabelText(useDefaultImage).click()
      expect(queryByRole('alert')).toBeFalsy()
    })

    test('Can go to next page (upload image, switch to default) - used to be a bug', () => {
      const { getByText, getByLabelText } = renderWithState(IMAGE_SELECTED_FOR_UPLOAD)
      getByLabelText(useDefaultImage).click()
      getByText('Redigera text').click()
      expect(getByText('Granska')).toBeInTheDocument()
    })

    test('Has correct alert text (published image, upload image, switch to default)', () => {
      const { getByRole, getByLabelText, getByTestId } = renderWithState(OVERWRITE_PUBLISHED_IMAGE)
      getByLabelText(useDefaultImage).click()
      expect(getByRole('alert')).toHaveTextContent(expected)
    })
  })

  describe('Page 1H Godkänna villkor', () => {
    beforeAll(() => mockClientFunctionsToWebContext())
    afterAll(() => {
      jest.clearAllMocks()
    })

    const expected =
      'Du behöver godkänna villkoren (se markering i rött nedan) för att kunna gå vidare till ”Redigera text”.'

    test('Has correct alert text (selected image, not agreed to terms)', () => {
      const { getByRole, getByTestId } = renderWithState(IMAGE_SELECTED_FOR_UPLOAD)
      getByTestId('termsAgreement').click()
      getByTestId('termsAgreement').click()
      expect(getByRole('alert')).toHaveTextContent(expected)
    })

    test('Must tick box to continue if image was uploaded', () => {
      const { getByLabelText, getByTestId, getByText, getByRole, queryByRole, queryByTestId } =
        renderWithState(IMAGE_SELECTED_FOR_UPLOAD)
      expect(getByLabelText('Standardbild utifrån kursens huvudområde').checked).toBeFalsy()
      expect(getByLabelText('Egen bild').checked).toBeTruthy()

      //No error message visible initially
      expect(queryByTestId('error-text')).toBeFalsy()

      getByText('Redigera text').click()

      // Expect alert, error message and a disabled button
      expect(getByRole('alert')).toHaveTextContent(expected)
      const errorText = getByTestId('error-text')
      expect(errorText).toHaveClass('error-label')
      expect(getByText('Redigera text').disabled).toBeTruthy()

      getByTestId('termsAgreement').click()

      // Expect no alert, no error message and an enabled button
      expect(queryByRole('alert')).toBeFalsy()
      expect(queryByTestId('error-text')).toBeFalsy()
      expect(getByText('Redigera text').disabled).toBeFalsy()
    })
  })

  describe('Page 2 Redigera introduktion till kursen', () => {
    beforeAll(() => mockClientFunctionsToWebContext())
    afterAll(() => {
      jest.clearAllMocks()
    })

    test('Has correct introductory text', () => {
      const pageNumber = 2
      const { getByTestId } = renderEditPage({}, pageNumber)
      const introText = getByTestId('intro-text')
      expect(introText).toHaveTextContent('Här lägger du in en text som beskriver kursen.')
      expect(introText).toHaveTextContent('Texten kommer att visas för din kurs på sidan Inför kursval.')
      expect(introText).toHaveTextContent(
        'Det kan finnas en beskrivande text inlagd sedan tidigare via Kopps/Ladok, men om du lägger in en text här är det den som visas på sidan Inför kursval.'
      )
    })
  })

  describe('Page 3 Granska', () => {
    beforeAll(() => mockClientFunctionsToWebContext())

    afterAll(() => {
      jest.clearAllMocks()
    })

    const pageNumber = 3

    test('Has correct introductory text', () => {
      const introText = renderEditPage({}, pageNumber).getByTestId('intro-text')
      expect(introText).toHaveTextContent('')
    })

    test('Has correct headings', () => {
      const { getByText } = renderEditPage({}, pageNumber)
      getByText('Svensk introduktion till kursen')
      getByText('Engelsk introduktion till kursen')
    })
  })

  describe('Page 3B Att tänka på innan du publicerar', () => {
    beforeAll(() => mockClientFunctionsToWebContext())

    afterAll(() => {
      jest.clearAllMocks()
    })

    const pageNumber = 3

    test('Has correct modal text', async () => {
      const { getByText } = renderWithState({}, pageNumber)
      getByText('Publicera').click()
      expect(getByText(/Kurs: SF1624/)).toBeTruthy()
      const regExpected = /Publicering kommer att ske på sidan ”Inför kursval”\./
      expect(getByText(regExpected)).toBeTruthy()
      expect(getByText(/Vill du fortsätta att publicera?/)).toBeTruthy()
    })
  })

  describe('Page 3C Publicering fel', () => {
    beforeAll(() => mockClientFunctionsToWebContext())
    afterAll(() => {
      jest.clearAllMocks()
    })

    const pageNumber = 3

    test('Has correct alert text', async () => {
      const oldXMLHttpRequest = window.XMLHttpRequest
      window.XMLHttpRequest = jest.fn(() => {
        return {
          upload: {
            addEventListener: jest.fn(() => {
              throw 'trigger error state, for test purposes'
            }),
          },
        }
      })

      const { getByText, findByRole, getByRole } = await renderWithState(IMAGE_SELECTED_FOR_UPLOAD, pageNumber)
      fireEvent.click(getByText('Publicera'))
      await fireEvent.click(getByText('Ja, fortsätt publicera'))
      await waitFor(() => {
        const alert = getByRole('alert')

        expect(alert).toHaveTextContent(
          'Det gick inte att publicera den bild du valt. Gå tillbaka till ”Välj bild” för att byta bild. Prova sedan att ”Publicera”.'
        )
      })

      window.XMLHttpRequest = oldXMLHttpRequest
    })
  })

  describe('Page 3 Publish Progress Bar', () => {
    beforeAll(() => mockClientFunctionsToWebContext())
    afterAll(() => {
      jest.clearAllMocks()
    })

    test('Progress bar should not be visible before publish action', () => {
      const pageNumber = 3
      const { queryByRole, getByText, findByRole } = renderEditPage({}, pageNumber)
      expect(queryByRole('status')).toBeNull()

      getByText('Publicera').click()
      getByText('Ja, fortsätt publicera').click()
      expect(findByRole('status')).toBeTruthy()
    })
  })
})
