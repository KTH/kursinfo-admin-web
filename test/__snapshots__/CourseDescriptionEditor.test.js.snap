// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CourseDescriptionEditorPage> (and subordinates) Renders without errors (incl. snapshot) 1`] = `
<DocumentFragment>
  <div
    class="kursinfo-main-page col"
  >
    <header
      class="pageTitle col"
      id="course-title"
    >
      <h1
        data-testid="main-heading"
      >
        Redigera introduktion till kursen
      </h1>
      <div
        data-testid="main-course"
      >
        <span
          property="aiiso:code"
        >
          SF1624
        </span>
        <span
          property="teach:pageTitle"
        >
           Algebra och geometri
        </span>
        <span
          content="7.5"
          datatype="xsd:decimal"
          property="teach:ects"
        >
           7,5 hp
        </span>
      </div>
    </header>
    <div
      class="progress-bar-container row"
    >
      <div
        class="col-md-4 col-sm-12 progress-bar1 progress-active"
      >
        <span>
          1. Välj bild
        </span>
      </div>
      <div
        class="col-md-4 col-sm-12 progress-bar1 "
      >
        <span>
          2. Redigera text
        </span>
      </div>
      <div
        class="col-md-4 col-sm-12 progress-bar1 "
      >
        <span>
          3. Granska och publicera
        </span>
      </div>
      <div>
        <p
          data-testid="intro-text"
        >
          Börja med att välja vilken bild som ska visas på kursinformationssidan (steg 1 av 3). I nästa steg (2 av 3) kommer du att kunna redigera den inledande texten. 
    I sista steget (3 av 3) ges möjlighet att först granska bild och text och sedan publicera det på sidan ”Kursinformation”.
        </p>
      </div>
    </div>
    <span
      class="Upload--Area col"
    >
      <span
        class="title_and_info"
      >
        <h2
          data-testid="intro-heading"
        >
          Välj bild
          <span>
            <button
              class="btn-info-modal btn btn-btn-info-modal"
              type="button"
            />
          </span>
        </h2>
      </span>
      <p>
        Välj bild som ska visas på kursinformationssidan:
      </p>
      <form
        class="Picture--Options input-label-row"
      >
        <span
          role="radiogroup"
        >
          <label
            for="defaultPicture"
          >
            <input
              checked=""
              id="defaultPicture"
              name="choosePicture"
              type="radio"
              value="defaultPicture"
            />
             Bild vald utifrån kursens huvudområde
          </label>
           
          <br />
          <label
            for="otherPicture"
          >
            <input
              id="otherPicture"
              name="choosePicture"
              type="radio"
              value="otherPicture"
            />
             Egen vald bild
          </label>
           
          <br />
        </span>
      </form>
      <span
        class=""
      >
        <img
          alt="Bild för kurssidasdekoration"
          height="auto"
          src="Picture_by_MainFieldOfStudy_11_Mathematics.jpg"
          width="300px"
        />
      </span>
      <span
        class="control-buttons"
      >
        <div
          class="col-sm-4"
        />
        <div
          class="btn-cancel col-sm-4"
        >
          <span>
            <button
              class="secondary btn btn-secondary"
              type="button"
            >
              Avbryt
            </button>
          </span>
        </div>
        <div
          class="step-forward col-sm-4"
        >
          <button
            alt="Till nästa steg att redigera text"
            class="btn-next btn btn-success"
            type="button"
          >
            Redigera text
          </button>
        </div>
      </span>
    </span>
  </div>
</DocumentFragment>
`;
